### PR TITLE
Documentation: reorganize home page per comments

### PR DESCRIPTION
**Pull request by davidcok**
Edits to and reorganization of the documentation home page, per comments.

### CODE REVIEW

File: docs/QuickReference.html [hunk 1]
```diff
@@ -2,92 +2,216 @@
...
 <p>The assignment statement assigns each right-hand side in ExprList to the corresponding left-hand side in Lvalues. These assignments are performed in parallel (more to the point, all necessary reads occur before the writes), so the left-hand sides must denote distinct L-values. Each right-hand side can be an expression or an object creation of one of the following forms:</p>
-<p>new Tnew T.Init(ExprList)new T(ExprList)new T[SizeExpr]new T[SizeExpr0, SizeExpr1]</p>
-<p>The first form allocates an object of type T. The second form additionally invokes an initialization method or constructor on the newly allocated object. The third form shows how the syntax differs from the second form when the constructor called is anonymous.  The other forms show examples of array allocations, in particular a one- and a two-dimensional array of T values, respectively.</p>
+<pre>
+  new Tnew T.Init(ExprList)
```
Comment (RustanLeino) [inline docs/QuickReference.html:164]: break into two lines

Comment (RustanLeino) [reply docs/QuickReference.html:164]: Add an example that shows the syntax when the constructor is anonymous:

    new T(ExprList)

File: docs/QuickReference.html [hunk 2]
```diff
@@ -2,92 +2,216 @@
...
+
+<p>Operations on sequences include + (concatenation) and the comparison operators < (proper prefix) and <= (prefix). 
+Membership can be checked like for sets: <code>x in S</code> and <code>x !in S</code>. 
+The length of a sequence S is denoted <code>|S|</code>, and the elements of such a sequence have indices from 0 to less than <code>|S|</code>. 
+The expression <code>S[j]</code> denotes the element at index j of sequence S. 
+The expression <code>S[m..n]</code>, where 0 <= m <= n <= |S|, returns a sequence whose elements are the n-m elements of S starting at index m (that is, from S&#091;m&#093;, S&#091;m+1&#093;, ... up to but not including S&#091;n&#093;). 
+The expression S&#091;m..&#093; (often called "drop m") is the same as S&#091;m..|S|&#093;, that is, it returns the sequence whose elements are all but the first m elements of S. 
+The expression S&#091;..n&#093; (often called "take n") is the same as S&#091;0..n&#093;, that is, it returns the sequence that consists of the first n elements of S.
```
Comment (RustanLeino) [inline docs/QuickReference.html:205]: These should be formatted like in the previous lines

File: docs/QuickReference.html [hunk 3]
```diff
@@ -2,92 +2,216 @@
...
+<p>A method can be declared as ghost by preceding the declaration with the keyword ghost. By default, a method in a class has an implicit receiver parameter, this. This parameter can be removed by preceding the method declaration with the keyword static. A static method M in a class C can be invoked by C.M(...).</p>
+<p>In a class, a method can be declared to be a constructor method by replacing the keyword method with the keyword constructor. A constructor can only be called at the time an object is allocated (see object-creation examples below), and for a class that contains one or more constructors, object creation must be done in conjunction with a call to a constructor.  Ordinarily, a method has a name, of course, but a class is allowed to have one constructor without a name&#8211;an <em>anonymous constructor</em>&#8211;like this:</p>
+<pre>
+  constructor (n: int)
+    modifies this 
+  {
+    Body
+  }
+</pre>
+<p>Often, ghost methods are used as lemmas.  This can be made clearer in the program text by declaring the method with lemma instead of ghost method.</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:44]: Reworded sentence. Also rewrote entire document as .md and added in backticks as appropriate.

File: docs/QuickReference.html [hunk 4]
```diff
@@ -2,92 +2,216 @@
...
-<p>modifies {c, d}</p>
-<p>modifies {c} + {d}</p>
-<p>modifies c, {d}</p>
-<p>modifies c, d</p>
+<pre>
+method M(a: A, b: B, c: C) returns (x: X, y: Y, z: Y)
+  requires Pre 
+  modifies Frame 
+  ensures Post 
+  decreases Rank
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:20]: Done - and elsewhere

File: docs/QuickReference.html [hunk 5]
```diff
@@ -2,92 +2,216 @@
...
-<p>In a class, a method can be declared to be a constructor method by replacing the keyword method with the keyword constructor. A constructor can only be called at the time an object is allocated (see object-creation examples below), and for a class that contains one or more constructors, object creation must be done in conjunction with a call to a constructor.  Ordinarily, a method has a name, of course, but a class is allowed to have one constructor without a name&#8211;an <em>anonymous constructor</em>&#8211;like this:</p>
-<p>constructor (n: int)modifies this{Body}</p>
-<p>Often, ghost methods are used as lemmas.  This can be made clearer in the program text by declaring the method with lemma instead of ghost method.</p>
+<p>If omitted, the pre- and postconditions default to true and the frame defaults to the empty set. The variant function is a list of expressions, denoting the unending lexicographic tuple consisting of the given expressions followed implicitly by “top” elements. If omitted, Dafny will guess a variant function for the method, namely the lexicographic tuple that starts with the list of the method's in-parameters.</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:34]: Done

File: docs/QuickReference.html [hunk 6]
```diff
@@ -2,92 +2,216 @@
...
+<pre>
+function Factorial(n: int): int
+  requires 0 <= n
+ensures 1 <= Factorial(n)
+{
+if n == 0 then 1 else Factorial(n-1) * n
+}
+</pre>
+<p>says that the result of Factorial is always positive, which Dafny verifies inductively from the function body. To refer to the function's result in the postcondition, use the function itself, as shown in the example.</p>
+<p>By default, a function is ghost, and cannot be called from non-ghost code. To make it non-ghost, replace the keyword function with the two keywords function method.  A function that returns a boolean can be declared with the keyword predicate (and then eliding the colon and return type).</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:90]: Done

File: docs/QuickReference.html [hunk 7]
```diff
@@ -2,92 +2,216 @@
...
 <p>A constructor can optionally declare a destructor for any of its parameters, which is done by introducing a name for the parameter. For example, if Tree were declared as:</p>
-<p>datatype Tree = Leaf | Node(left: Tree, data: int, right: Tree)</p>
-<p>then t1.data == 5 and t1.left == t0 hold after the code snippet above.</p>
+<pre>datatype Tree = Leaf | Node(left: Tree, data: int, right: Tree)</pre>
+<p>then <code>t1.data == 5</code> and <code>t1.left == t0</code> hold after the code snippet above.</p>
 <h2>Generics</h2>
-<p>Dafny supports type parameters.  That is, any class, inductive datatype, method, and function can have type parameters. These are declared in angle brackets after the name of what is being declared. For example:</p>
-<p>class MyMultiset<t> { /*…*/ }datatype Tree</t><t> = Leaf | Node(Tree</t><t>, T, Tree</t><t>)method Find</t><t>(key: T, collection: Tree</t><t>) { /*…*/ }function IfThenElse</t><t>(b: bool, x: T, y: T): T { /*…*/ }</t></p>
+<p>Dafny supports type parameters.  That is, any class, inductive datatype, method, and function can have type parameters. These are declared in angle brackets after the name of what is being declared. For example:</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:112]: Done

File: docs/QuickReference.html [hunk 8]
```diff
@@ -2,92 +2,216 @@
...
+<pre>
+class MyMultiset<t> {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:114]: Fixed, and switched to md

File: docs/QuickReference.html [hunk 9]
```diff
@@ -2,92 +2,216 @@
...
-<p>if BoolExpr0 {Stmts0} else if BoolExpr1 {Stmts1} else {Stmts2}</p>
-<p>while BoolExprinvariant Invmodifies Framedecreases Rank{Stmts}</p>
-<p>match Expr {case Empty => Stmts0case Node(l, d, r) => Stmts1}</p>
-<p>break;return;</p>
+<pre>
+  var LocalVariables := ExprList;
+
+  Lvalues := ExprList;
+
+  assert BoolExpr;print ExprList;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:134]: Done

File: docs/QuickReference.html [hunk 10]
```diff
@@ -2,92 +2,216 @@
...
+Post is a boolean expression denoting the function's postcondition, 
+Rank is the function's variant function, 
+and Body is an expression that defines the function.  
+The precondition allows a function to be partial, that is, the precondition says when the function is defined (and Dafny will verify that every use of the function meets the precondition). The postcondition is usually not needed, since the body of the function gives the full definition. However, the postcondition can be a convenient place to declare properties of the function that may require an inductive proof to establish. For example:</p>
+<pre>
+function Factorial(n: int): int
+  requires 0 <= n
+ensures 1 <= Factorial(n)
+{
+if n == 0 then 1 else Factorial(n-1) * n
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:86]: Done

File: docs/QuickReference.html [hunk 11]
```diff
@@ -2,92 +2,216 @@
...
+}
+</pre>
+<p>says that the result of Factorial is always positive, which Dafny verifies inductively from the function body. To refer to the function's result in the postcondition, use the function itself, as shown in the example.</p>
+<p>By default, a function is ghost, and cannot be called from non-ghost code. To make it non-ghost, replace the keyword function with the two keywords function method.  A function that returns a boolean can be declared with the keyword predicate (and then eliding the colon and return type).</p>
+<p>By default, a function in a class has an implicit receiver parameter, this. This parameter can be removed by preceding the function declaration with the keyword static. A static function F in a class C can be invoked by F.M(...). This can give a convenient way to declare a number of helper functions in a separate class.</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:91]: Done

File: docs/QuickReference.html [hunk 12]
```diff
@@ -2,92 +2,216 @@
...
+<pre>class C {
+// member declarations go here
+}
+</pre>
 <p>where the members of the class (fields, methods, and functions) are defined (as described above) inside the curly braces.</p>
 <h2>Datatypes</h2>
 <p>An inductive datatype is a type whose values are created using a fixed set of constructors. A datatype Tree with constructors Leaf and Node is declared as follows:</p>
-<p>datatype Tree = Leaf | Node(Tree, int, Tree)</p>
+<pre>datatype Tree = Leaf | Node(Tree, int, Tree)</pre>
 <p>The constructors are separated by vertical bars. Parameter-less constructors need not use parentheses, as is shown here for Leaf.</p>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/QuickReference.html:102]: Reworded

File: docs/QuickReference.html (hunk 1 above)
Comment (davidcok) [reply docs/QuickReference.html:164]: Done

File: docs/QuickReference.html (hunk 2 above)
Comment (davidcok) [reply docs/QuickReference.html:205]: Done

**Pull request merged**